Sources/SystemInternals/Mocking.swift

```diff
@@ -0,0 +1,164 @@
+
+  public mutating func dequeue() -> Entry? {
+    guard !self.isEmpty else { return nil }
+    defer { firstEntry += 1 }
+    return entries[firstEntry]
+  }
+
+  internal mutating func add(_ e: Entry) {
+    entries.append(e)
+  }
+
+  public mutating func clear() { entries.removeAll() }
+}
+
+// TODO: Track
+public struct WriteBuffer {
+  public var enabled: Bool = false
+
+  private var buffer: [UInt8] = []
+  private var chunkSize: Int? = nil
+
+  internal mutating func write(_ buf: UnsafeRawBufferPointer) -> Int {
+    guard enabled else { return 0 }
+    let chunk = chunkSize ?? buf.count
+    buffer.append(contentsOf: buf.prefix(chunk))
+    return chunk
+  }
+
+  public var contents: [UInt8] { buffer }
+}
+
+public enum ForceErrno: Equatable {
+  case none
+  case always(errno: CInt)
+
+  case counted(errno: CInt, count: Int)
+}
+
+// Provide access to the driver, context, and trace stack of mocking
+public class MockingDriver {
+  // Record syscalls and their arguments
+  public var trace = Trace()
+
+  // Mock errors inside syscalls
+  public var forceErrno = ForceErrno.none
+
+  // A buffer to put `write` bytes into
+  public var writeBuffer = WriteBuffer()
+}
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import Darwin
+#elseif os(Linux) || os(FreeBSD) || os(Android)
+import Glibc
+#else
+#error("Unsupported Platform")
+#endif
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+private func releaseObject(_ raw: UnsafeMutableRawPointer) -> () {
+  Unmanaged<MockingDriver>.fromOpaque(raw).release()
+}
+#elseif os(Linux) || os(FreeBSD) || os(Android)
+private func releaseObject(_ raw: UnsafeMutableRawPointer?) -> () {
+  guard let object = raw else { return }
+  Unmanaged<MockingDriver>.fromOpaque(object).release()
+}
+#else
+#error("Unsupported Platform")
+#endif
+
+
+internal let key: pthread_key_t = {
+  var raw = pthread_key_t()
+  guard 0 == pthread_key_create(&raw, releaseObject) else {
+    fatalError("Unable to create key")
+  }
+  return raw
+}()
+
+internal var currentMockingDriver: MockingDriver? {
+  #if !ENABLE_MOCKING
+    fatalError("Contextual mocking in non-mocking build")
+  #endif
+
+  guard let rawPtr = pthread_getspecific(key) else { return nil }
+
+  return Unmanaged<MockingDriver>.fromOpaque(rawPtr).takeUnretainedValue()
+}
+
+extension MockingDriver {
+  /// Enables mocking for the duration of `f` with a clean trace queue
+  /// Restores prior mocking status and trace queue after execution
+  public static func withMockingEnabled(
+    _ f: (MockingDriver) throws -> ()
+  ) rethrows {
+    let priorMocking = currentMockingDriver
+    defer {
+      if let object = priorMocking {
+        pthread_setspecific(key, Unmanaged.passUnretained(object).toOpaque())
+      } else {
+        pthread_setspecific(key, nil)
+      }
+    }
+
+    let driver = MockingDriver()
+    guard 0 == pthread_setspecific(key, Unmanaged.passRetained(driver).toOpaque()) else {
```

Comment:
Hmm, I think we should pass it unretained, fix the lifetime, and remove the `releaseObject` destructor above.

Currently, if the thread is killed and the TLS entry is non-NULL, `releaseObject` is called. If this is the top-most `withMockingEnabled`, then the entry will be set to NULL and there will be a leak. If this is a nested call, then the parent driver will be restored (AFAICT this is balanced), and this will leak. If the thread is killed during this call, then the "passRetained" will be paired with the the release in `releaseObject`, IIUC leakiness depends on whether there are other references to the driver alive.

But, if we pass it unretained and fix the lifetime, it will be balanced for the duration of the function and the `defer` above will remove the TLS entry anyways. If we don't have a `releaseObject` registered, then our leaking story if the thread is killed during this function is just the normal one for Swift.

Reply:
My point is that the destructor hook isn't called by `pthread_setspecific` -- it is only called at thread exit. When `withMockingEnabled` restores the original driver, it must also release the driver that it replaced. 

(It doesn't seem feasible to prevent all leaks if someone calls `pthread_exit`, but we should make sure not to leak in regular use.)

Slight nit: it would make this code slightly simpler if it simply used `getspecific` instead of `currentMockingDriver` -- it wouldn't need to deal with the back-n-forth conversion and optionals at all.

Reply:
Ah, I see what you mean -- removing the destructor could work too!